images/builder/Dockerfile

```diff
@@ -0,0 +1,128 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#######################################################################
+###
+### GOLANG BUILDER - For building any go binaries needed in later steps
+###
+#######################################################################
+
+FROM golang:1.12 AS builder
+COPY barnacle/* $GOPATH/src/barnacle/
+
+# RUN apt-get install -y --no-install-recommends ca-certificates git
+RUN go get -u github.com/golang/dep/cmd/dep
+WORKDIR $GOPATH/src/barnacle
+RUN dep ensure --vendor-only
+RUN go build -o /barnacle $GOPATH/src/barnacle
+
+##################################################
+###
+### Builder image including Docker-In-Docker setup
+###
+##################################################
+
+FROM debian:stretch
+
+WORKDIR /workspace
+RUN mkdir -p /workspace
+ENV WORKSPACE=/workspace \
+    TERM=xterm
+
+# add env we can debug with the image name:tag
+ARG IMAGE_ARG
+ENV IMAGE=${IMAGE_ARG}
+
+# common util tools
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    build-essential \
+    ca-certificates \
+    curl \
+    file \
+    git \
+    jq \
+    mercurial \
+    openssh-client \
+    pkg-config \
+    procps \
+    rsync \
+    unzip \
+    wget \
+    xz-utils \
+    zip \
+    zlib1g-dev \
+    && rm -rf /var/lib/apt/lists/*
+
+#
+# BEGIN: DOCKER IN DOCKER SETUP
+#
+
+# Install Docker deps, some of these are already installed in the image but
+# that's fine since they won't re-install and we can reuse the code below
+# for another image someday.
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    apt-transport-https \
+    ca-certificates \
+    curl \
+    gnupg2 \
+    software-properties-common \
+    lsb-release && \
+    rm -rf /var/lib/apt/lists/*
+
+# Add the Docker apt-repository
+RUN curl -fsSL https://download.docker.com/linux/$(. /etc/os-release; echo "$ID")/gpg \
+    | apt-key add - && \
+    add-apt-repository \
+    "deb [arch=amd64] https://download.docker.com/linux/$(. /etc/os-release; echo "$ID") \
+    $(lsb_release -cs) stable"
+
+# Install Docker
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends docker-ce=18.06.* && \
+    rm -rf /var/lib/apt/lists/* && \
+    sed -i 's/cgroupfs_mount$/#cgroupfs_mount\n/' /etc/init.d/docker
+
+
+# Move Docker's storage location
+RUN echo 'DOCKER_OPTS="${DOCKER_OPTS} --data-root=/docker-graph"' | \
+    tee --append /etc/default/docker
+# NOTE this should be mounted and persisted as a volume ideally (!)
+# We will make a fallback one now just in case
+RUN mkdir /docker-graph
+
+# add custom docker cleanup binary
+COPY --from=builder ["/barnacle", "/usr/local/bin/"]
+
+#
+# END: DOCKER IN DOCKER SETUP
+#
+
+# Run actual from here on out as the prow user
```

Comment:
Missed word?
```suggestion
# Run actual docker-in-docker from here on out as the prow user
```

Reply:
I agree the sentence needs a word but not sure I want to put `docker-in-docker`, this is more about the future instructions and actual CI pipelines rather than specifically DinD, perhaps:
```suggestion
# Run everything from here on out as the prow user
```

WDYT?

Reply:
That's good.